flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/CommittableToTableChangeConverter.java

```diff
@@ -0,0 +1,178 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.flink.sink;
+
+import java.io.IOException;
+import java.util.List;
+import org.apache.flink.api.common.state.ListState;
+import org.apache.flink.api.common.state.ListStateDescriptor;
+import org.apache.flink.core.io.SimpleVersionedSerialization;
+import org.apache.flink.runtime.state.StateInitializationContext;
+import org.apache.flink.runtime.state.StateSnapshotContext;
+import org.apache.flink.streaming.api.connector.sink2.CommittableMessage;
+import org.apache.flink.streaming.api.connector.sink2.CommittableWithLineage;
+import org.apache.flink.streaming.api.operators.AbstractStreamOperator;
+import org.apache.flink.streaming.api.operators.OneInputStreamOperator;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.iceberg.DataFile;
+import org.apache.iceberg.DeleteFile;
+import org.apache.iceberg.ManifestFile;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.flink.TableLoader;
+import org.apache.iceberg.flink.maintenance.operator.TableChange;
+import org.apache.iceberg.io.WriteResult;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CommittableToTableChangeConverter extends AbstractStreamOperator<TableChange>
+    implements OneInputStreamOperator<CommittableMessage<IcebergCommittable>, TableChange> {
+
+  private static final Logger LOG =
+      LoggerFactory.getLogger(CommittableToTableChangeConverter.class);
+
+  private final TableLoader tableLoader;
+  private Table table;
+  private transient ListState<ManifestFile> toRemoveManifestFileState;
+  private transient List<ManifestFile> toRemoveManifestFileList;
+  private transient long lastCompletedCheckpointId = -1L;
+  private transient String flinkJobId;
+
+  public CommittableToTableChangeConverter(TableLoader tableLoader) {
+    Preconditions.checkNotNull(tableLoader, "TableLoader should not be null");
+    this.tableLoader = tableLoader;
+  }
+
+  @Override
+  public void initializeState(StateInitializationContext context) throws Exception {
+    super.initializeState(context);
+    this.toRemoveManifestFileList = Lists.newArrayList();
+    this.flinkJobId = getContainingTask().getEnvironment().getJobID().toString();
+    this.toRemoveManifestFileState =
+        context
+            .getOperatorStateStore()
+            .getListState(new ListStateDescriptor<>("to-remove-manifest", ManifestFile.class));
+    if (context.isRestored()) {
+      toRemoveManifestFileList = Lists.newArrayList(toRemoveManifestFileState.get());
```

Comment:
should we call this `manifests-to-remove` and `manifestFilesToRemove` respectively?